lib/Dialect/Torch/Transforms/Passes.cpp

```diff
@@ -124,12 +124,14 @@ void mlir::torch::Torch::createTorchSimplificationPipeline(
   pm.addNestedPass<func::FuncOp>(Torch::createMaximizeValueSemanticsPass());
   // Update the return op to return value tensors.
   pm.addPass(Torch::createRefinePublicReturnPass());
-  pm.addNestedPass<func::FuncOp>(createCanonicalizerPass());
-  // Do shape refinement.
-  // This should be run before RefineTypes (which primarily does dtype
-  // inference), because Torch type promotion rules actually depend on the shape
-  // of the operand.
-  createTorchShapeRefinementPipeline(pm);
+  for (int i = 0; i < options.maxIterations; ++i) {
```

Comment:
maxIterations controls the iterations in LowerToBackendContractPass and should not be used this way.

Reply:
> createTorchSimplificationPipeline is already running iteratively inside LowerToBackendContractPass after #1165 Is this still an issue?

Yes. In my local development environment, if only run ShapeRefinementPipeline once, the output IR of the pipeline is something like this:
``` mlir
// -----// IR Dump After DropShapeCalculations (torch-drop-shape-calculations) //----- //
...
%5 = torch.vtensor.literal(dense<5> : tensor<si64>) : !torch.vtensor<[],si64>
%210 = torch.prim.NumToTensor.Scalar %int5 : !torch.int -> !torch.vtensor<[],unk>
%211 = torch.aten.div.Tensor_mode %210, %5, %str_4 : !torch.vtensor<[],unk>, !torch.vtensor<[],si64>, !torch.str -> !torch.vtensor<[],unk>
%212 = torch.aten.Int.Tensor %211 : !torch.vtensor<[],unk> -> !torch.int
%213 = torch.prim.ListConstruct %212, %int5, %int-1 : (!torch.int, !torch.int, !torch.int) -> !torch.list<int>
%214 = torch.aten.view %209, %213 : !torch.vtensor<[5,2048],unk>, !torch.list<int> -> !torch.vtensor<[?,5,?],unk>
...
```

Reply:
I think the issue is related to RefineTypes. Shall we add `PrimNumToTensorScalar` to https://github.com/llvm/torch-mlir/blob/bb47c166a047ddbc913950cb0f2571594c0fe261/lib/Dialect/Torch/Transforms/RefineTypes.cpp#L638?

Reply:
I think this issue is only about constant shape. Could you explain more on the RefineTypes's influence on this problem? In my experiments, the type inference for `PrimNumToTensorScalarOp` will fail if we do so.
```mlir
// -----// IR Dump After RefineTypes (torch-refine-types) //----- //
%210 = torch.prim.NumToTensor.Scalar %int5 : !torch.int -> !torch.vtensor<[],unk>
```